File: Textural-Feature-Extraction-main/Script/Python-CPU/GaborFilter.py
import numpy as np
from scipy.signal import convolve2d



def gabor_filter(I, Sx, Sy, Scale, Orientation):
    """
    Gabor Filter

    Args:
        I: Input image (original image)
        (Sx,Sy: control the size of the Gaussian envelope part of the Gabor function)
        Sx: Variances along the x-axis
        Sy: Variances along the y-axis
        Scale:scale parameter, control the filter's sensitivity to details
        Orientation: orientation parameter, defines the orientation of the Gabor filter

    Returns:
        G: Gabor Filter, a two-dimensional filter matrix
        gabout: Filtered Image Magnitude, the magnitude of the result
        gaboutReal: Real Part of Filtered Image, convolve the image with the cosine part of the Gabor filter
        gaboutImg: Imaginary Part of Filtered Image, convolve the image with the sine part of the Gabor filter
    """

    # Use float64 to ensure calculation accuracy
    if I.dtype != np.float64:
        I = I.astype(np.float64)

    m_dSigma = 2 * np.pi
    m_dfrequency = np.sqrt(2)
    m_dKmax = np.pi / 2

    HeightBottom = -int(Sy / 2)
    HeightTop = int(Sy / 2)

    WidthBottom = -int(Sy / 2)
    WidthTop = int(Sy / 2)

    postConstant = np.exp(-m_dSigma * m_dSigma / 2)

    preConstant_Kuv = (m_dKmax / m_dfrequency ** Scale) ** 2
    Phi = Orientation
    Kv = m_dKmax / (m_dfrequency ** Scale)

    m_dSigmaY = 2 * np.pi
    m_dSigmaX = 2 * np.pi

    G = np.zeros((WidthTop - WidthBottom + 1, HeightTop - HeightBottom + 1), dtype=np.complex_)  # np.complex128

    for x in range(WidthBottom, WidthTop + 1):
        for y in range(HeightBottom, HeightTop + 1):
            exppart = np.exp(-1 * preConstant_Kuv / 2 * (
                        x ** 2 / (m_dSigmaX ** 2) + y ** 2 / (m_dSigmaY ** 2))) * preConstant_Kuv / (
                                  m_dSigmaX * m_dSigmaY)
            G[x - WidthBottom, y - HeightBottom] = exppart * (
                        np.cos(Kv * (np.cos(Phi) * x + np.sin(Phi) * y)) - postConstant) + 1j * exppart * (
                                                       np.sin(Kv * (np.cos(Phi) * x + np.sin(Phi) * y)))

            # equivalent expression
            # G[WidthTop + x, HeightTop + y] = exppart * (
            #             np.cos(Kv * (np.cos(Phi) * x + np.sin(Phi) * y)) - postConstant) + 1j * exppart * np.sin(
            #     Kv * (np.cos(Phi) * x + np.sin(Phi) * y))


    Imgabout = convolve2d(I, np.imag(G), mode='same')
    Regabout = convolve2d(I, np.real(G), mode='same')

    gabout = np.sqrt(Imgabout ** 2 + Regabout ** 2)
    gaboutImg = Imgabout
    gaboutReal = Regabout

    return G, gabout, gaboutReal, gaboutImg

File: Textural-Feature-Extraction-main/Script/MATLAB-parallel/GaborFilter.m
%%%%%%%VERSION 2
%%ANOTHER DESCRIBTION OF GABOR FILTER

%The Gabor filter is basically a Gaussian (with variances sx and sy along x and y-axes respectively)
%modulated by a complex sinusoid (with centre frequencies U and V along x and y-axes respectively) 
%described by the following equation
%%
%                            -1     x' ^     y'  ^             
%%% G(x,y,theta,f) =  exp ([----{(----) 2+(----) 2}])*cos(2*pi*f*x');
%                             2    sx'       sy'
%%% x' = x*cos(theta)+y*sin(theta);
%%% y' = y*cos(theta)-x*sin(theta);

%% Describtion :

%% I : Input image
%% Sx & Sy : Variances along x and y-axes respectively
%% f : The frequency of the sinusoidal function
%% theta : The orientation of Gabor filter

%% G : The output filter as described above
%% gabout : The output filtered image



%%  Author : Ahmad poursaberi  e-mail : a.poursaberi@ece.ut.ac.ir
%%          Faulty of Engineering, Electrical&Computer Department,Tehran
%%          University,Iran,June 2004

function [G,gabout,gaboutReal,gaboutImg] = GaborFilter(I,Sx,Sy,Scale,Orientation);

if isa(I,'double')~=1 
    I = double(I);
end


m_dSigma  = 2*pi;
m_dfrequency = sqrt(2);
m_dKmax  = pi/2;

HeightBottom  = -fix(Sy/2);
HeightTop     = fix(Sy/2);

WidthBottom  = -fix(Sy/2);
WidthTop     = fix(Sy/2);

postConstant    = exp(-m_dSigma*m_dSigma/2);


preConstant_Kuv = (m_dKmax / m_dfrequency^Scale)^2;
Phi = Orientation;
Kv  = m_dKmax / (m_dfrequency^Scale);


m_dSigmaY = 2*pi;
m_dSigmaX = 2*pi;


for x = WidthBottom:WidthTop
    for y = HeightBottom:HeightTop 
        %xPrime = x * cos(theta) + y * sin(theta);
        %yPrime = y * cos(theta) - x * sin(theta);
        
        exppart = exp(-1 * preConstant_Kuv/2 * (x*x/(m_dSigmaX*m_dSigmaX) + y*y/(m_dSigmaY*m_dSigmaY))) * preConstant_Kuv / (m_dSigmaX*m_dSigmaY) ;
        %pImagereal[k] =  exppart * ( cos(  Kv*(  cos(Phi)*x + sin(Phi)*y ) ) - postConstant );
	    %pImageconj[k] =  exppart * ( sin(  Kv*(  cos(Phi)*x + sin(Phi)*y ) ) );
				
        
        G(WidthTop+x+1,HeightTop+y+1) = exppart * ( cos(  Kv*(  cos(Phi)*x + sin(Phi)*y ) ) - postConstant ) + i* exppart * ( sin(  Kv*(  cos(Phi)*x + sin(Phi)*y ) ) );
        
        %exp(-.5*((xPrime/Sx)^2+(yPrime/Sy)^2))*cos(2*pi*f*xPrime);
    end
end

Imgabout = conv2(I,double(imag(G)),'same');
Regabout = conv2(I,double(real(G)),'same');

gabout = sqrt(Imgabout.*Imgabout + Regabout.*Regabout);
gaboutImg = Imgabout;
gaboutReal = Regabout;